### PR TITLE
backup: distribute work evenly on all replicas

### LINKED ISSUE #114528: backupccl: follower reads not considered during distsql planning
During DistSQL planning the `followerReadOracle` is only picked under some conditions outlined over here - https://github.com/cockroachdb/cockroach/blob/cf5c2caa84c635acbc22e73c8f2db35aa9de6c71/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L196-L196. One of the conditions is `txn != nil`. Both backup and changefeed do not plan their DistSQL flows in the scope of a txn, and so pass in nil txns to `dsp. SetupAllNodesPlanningWithOracle`. This means that we always pick the `binPackingOracle` instead that prefers leaseholders - https://github.com/cockroachdb/cockroach/blob/3111a55e91dba9a88931fda725b767163c981ae7/pkg/sql/physicalplan/replicaoracle/oracle.go#L270.

When a backup is run with execution locality it constrains the execution and planning of the backup to only the nodes in the specified region. If the user has configured the backup to run in a non-primary region with follower replicas, we should be picking these replicas instead of reaching out to the leaseholders in the primary region. This way users can avoid backups interfering with foreground traffic. This is not likely today because of our inability to use the `followerReadOracle`.

The likely fix for this bug is to plumb a timestamp or txn from backup, cdc flow planning to where we can decide to use follower reads or not. 

Jira issue: CRDB-33547

### PR DESCRIPTION
**Pull request by dt**
This changes backup to use a new "BulkOracle" which randomly assigns spans to any replica, including followers (if follower reads are enabled), optional constrained to those matching the locality filter if one is in use. This is expected to be beneficial for bulk jobs that have large amounts of work that benefits from being spread evenly across nodes, and where planning on nodes that match the locality filter is more important than planning on a leaseholder or voter, particularly as by the time most reads will be done by the planned work they will be so backdated as to be able to be served by any replica.

Release note (bug fix): backups now evenly distribute work their work across all replicas, including followers, regardless of leaseholder placement.
Epic: none.

### CODE REVIEW

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (miretskiy) [inline pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:222]: small test for this function?

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (dt) [reply pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:208]: I called it "bulk" based on its usage rather than its behavior since I didn't want to be inaccurate about its behavior if I gave it behavior-based name. I worry I will be inaccurate if I just call it uniform or something, particularly since I hope to in the future teach this thing to be "sticky" to a given node _up to a point_ for adjacent spans.

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (dt) [reply pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:244]: I think that's a distsql level decision: it can and already does look at the node that this function picks for one range and then see if can or can't use that node, and keeps stats about what nodes are getting used and why. I don't really think it is worth re-doing that logic down here.

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (dt) [reply pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:238]: I suppose? but like... this is job planning -- compared to what we'll do _after_ planning this is peanuts. But sure, as long as this struct doesn't need to be threadsafe, if we're worried about perf here, I can reuse this slice, and also make it `[]int` of offsets into the replica slice to reduce its size.

Comment (dt) [reply pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:238]: actually no, I don't think I should worry about this allocating here. Right above this we're allocating a new slice every time in `kvcoord.NewReplicaSlice` so until this shows up in a profile I'm calling this premature optimization (and one which would make it non-thread safe).

**Review by yuzefovich — COMMENTED** [review]
Looks good to me. Probably should be backported too?

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-:-NjiXXem5BLUPIGZfo70:buc9uvo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, @msbutler, and @nvanbenschoten)

___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 248 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114710#gh-1399248952:-NjiWibS0ruGCtiryJQ8:b-5pkinn) ([raw file](https://github.com/cockroachdb/cockroach/blob/dacf778374f98e179b850052a580f2d951f1e81b/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L248)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

not sure if it's worth it.. but calling global rand many many times might be expensive enough to perhaps deserve the use of fastrand (from go runtime).
</blockquote></details>

I agree with Yevgeniy - we should avoid using the global rand in this function.

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (miretskiy) [reply pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:238]: the problem with this is that it will not show up in this functions profile; it will show up as gc runtime...
But we do know that this function will be called thousands of times -- why allocate more than we need to?
A quick micro benchmark will certainly show that.

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (dt) [reply pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:224]: Added the smallest of tests -- verifies that it doesn't return an error and when a loc filter is used it picks the right node. Doing anything else feels like it'd just be testing the rng though.

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (dt) [reply pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go]: Done

**Review by yuzefovich — COMMENTED** [review]
LGTM, but I'll defer to Yevgeniy for approval.

Also, did you mean to include the update to the contract of `replicaoracle.Oracle` into this PR?

> Reviewed 6 of 6 files at r4, 1 of 1 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-:-Njjxg3M2fVHcuUuWgy8:b-8c0sp6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, @msbutler, and @nvanbenschoten)

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (dt) [reply pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go]: The existing oracle -- which we're calling thousands of times just like this one -- calls `replicaSliceOrErr` for every span which allocates a whole ` kvcoord.NewReplicaSlice` every call. If we want to start golfing allocations in this thing -- including the existing implementations as well -- then fine we can do that, but let's do it in an isolated performance-change-only PR, that has its own backport discussion, not in a behavior change.

**Comment by dt** [comment]
> Also, did you mean to include the update to the contract of replicaoracle.Oracle into this PR?

I removed the statefulness from the struct so I no longer depend on that contract change here. If/when we introduce something like allocation-reducing reusable scratch buffers, we can stipulate it, but for now I wanted to keep the diff minimal as dictated by the backport policy.

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (miretskiy) [reply pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go]: Sounds good.  Memoizing allocations does not make sense when you're doing one off queries; it's just seeing what
trying to plan a flow w/ 2.5M ranges would do, that made me suggest it.  Totally fine in a standalone PR.

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (miretskiy) [inline pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:231]: if leaseholder is nil and follower reads disabled, we will panic.
I don't think that's what we want.

File: pkg/sql/distsql_physical_planner.go
Comment (miretskiy) [inline pkg/sql/distsql_physical_planner.go:262]: returns oracle config?

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
Comment (miretskiy) [inline pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go:711]: let's add "follower reads disabled" test.

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
Comment (miretskiy) [inline pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go:709]: let's check that we at least got someting out of this call (like some sort of descriptor with a non-zero field (like node id or whatnot).

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
Comment (miretskiy) [inline pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go:713]: let's add a test that tries to choose prefer replica when no locality match exists.

**Review by miretskiy — APPROVED** [review]
:lgtm: modulo few suggestions for the test.

**Review by nvb — COMMENTED** [review]
This approach makes sense to me. We're randomly picking a replica for each range, so nodes that have more replicas will be given more work than nodes with fewer replicas. We could try to balance work more evenly even if replica distribution is uneven, but it's not clear that doing so would be desirable, so better to keep it simple.

> Reviewed 2 of 6 files at r4, 4 of 4 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-:-NjoCiUt2TXRJoZnUB-d:bces0xa)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @miretskiy, and @msbutler)

___
*[`pkg/ccl/backupccl/backup_job.go` line 210 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-NjoBvUu0RrPwAoHlCfX:-NjoBvUu0RrPwAoHlCfY:bs87tb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b2af5b4ee5f349f5230f0d849988969f3996eba/pkg/ccl/backupccl/backup_job.go#L210)):*
> ```Go
> 
> 	oracle := physicalplan.DefaultReplicaChooser
> 	if useBulkOracle.Get(&evalCtx.Settings.SV) {
> ```

In https://github.com/cockroachdb/cockroach/pull/114532#pullrequestreview-1737310211, we were talking about non-`AS OF SYSTEM TIME` backups and how they wouldn't be planned to take advantage of follower reads. Is this PR going in the opposite direction and planning on followers regardless of `AS OF SYSTEM TIME`? If we don't do anything else, will this lead to a few seconds of non-local Export requests until follower reads kick in?
___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 238 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114710#gh-1399233601:-NjoBGe5Ec2BnALP2s1y:blu41nn) ([raw file](https://github.com/cockroachdb/cockroach/blob/dacf778374f98e179b850052a580f2d951f1e81b/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L238)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Sounds good.  Memoizing allocations does not make sense when you're doing one off queries; it's just seeing what
trying to plan a flow w/ 2.5M ranges would do, that made me suggest it.  Totally fine in a standalone PR.
</blockquote></details>

We could use a `util.FastIntMap` to keep this on the stack.
___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 216 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-Njo9vi08l_QpSFssYjP:-Njo9vi08l_QpSFssYjQ:b-kl5hj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/de574fcf1cf1e340a369c3956190947535054dec/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L216)):*
> ```Go
> 
> // NewBulkOracle returns an oracle for planning bulk operations, which will plan
> // balancing random across all replicas if follower reads are enabled.
> ```

"will plan balancing random" sounds off. "randomly"?
___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 208 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-NjoDaB8Asij1Yvg0X3x:-NjoDaB8Asij1Yvg0X3y:b-jh4mfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b2af5b4ee5f349f5230f0d849988969f3996eba/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L208)):*
> ```Go
> var followerReadOraclePolicy = replicaoracle.RegisterPolicy(newFollowerReadOracle)
> 
> type bulkOracle struct {
> ```

Should this live in `pkg/sql/physicalplan/replicaoracle/oracle.go` next to the rest of the `Oracle` implementations?
___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 245 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-NjoDzqw9rCTSMrKV_lg:-NjoDzqw9rCTSMrKV_lh:b-4a9ze1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b2af5b4ee5f349f5230f0d849988969f3996eba/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L245)):*
> ```Go
> 		}
> 		if len(matches) > 0 {
> 			return replicas[matches[randutil.FastUint32()%uint32(len(matches))]].ReplicaDescriptor, false, nil
> ```

Should we be returning `true` for `ignoreMisplannedRanges` here and below? For the same reason that `closestOracle.ChoosePreferredReplica` does.

**Comment by miretskiy** [comment]
> This approach makes sense to me. We're randomly picking a replica for each range, so nodes that have more replicas will be given more work than nodes with fewer replicas. We could try to balance work more evenly even if replica distribution is uneven, but it's not clear that doing so would be desirable, so better to keep it simple.

I believe @dt has a plan for this in a follow up PR.

**Review by dt — COMMENTED** [review]
>  nodes that have more replicas will be given more work
This feels like a feature rather than a bug to me? Particularly if we consider heterogeneous nodes. A node that has more replicas should already expect to do more work than one that has fewer imo, so having backup/bulk work mirror that distribution seems good to me.

My plan for a followup PR is not to balance the work differently but to allocate the balanced work differently, specifically to try to assign adjacent spans to the same node when we can do so without unbalancing the allocation. But that is out of scope for now, and I think work distribution that matches replica distribution seems like a good property for now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-:-Njomb0TArLLPv4OnXHQ:b-bj91to)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @miretskiy, @msbutler, and @nvanbenschoten)

___
*[`pkg/ccl/backupccl/backup_job.go` line 210 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-NjoBvUu0RrPwAoHlCfX:-Njos8o6C5Ul4EcXnCAx:bres7bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b2af5b4ee5f349f5230f0d849988969f3996eba/pkg/ccl/backupccl/backup_job.go#L210)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

In https://github.com/cockroachdb/cockroach/pull/114532#pullrequestreview-1737310211, we were talking about non-`AS OF SYSTEM TIME` backups and how they wouldn't be planned to take advantage of follower reads. Is this PR going in the opposite direction and planning on followers regardless of `AS OF SYSTEM TIME`? If we don't do anything else, will this lead to a few seconds of non-local Export requests until follower reads kick in?
</blockquote></details>

I'm unconcerned about this, for two reasons.

First, as you point out here and in 14532, the backup/bulk job will run for minutes, and the overwhelming majority of that time the read timestamp will be even more in the past than it is during planning. So we should plan for what's optimal for the majority of requests. 

Second: backups are typically planned with a read timestamp already at least a couple seconds in the past: our docs advice using `-10s` or more for manual backups, and scheduled backups run with their AOST set to the _scheduled_ time, which is always prior to the execution time since a schedule is only executed once a node's periodic poll happens to notice a schedule's planned execution time is in the past. 

Backups already benefit from being backdated by some number of seconds w.r.t. contention and intents, so if this is yet another reason to backdate them, that seems a-ok to me.
___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 238 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114710#gh-1399233601:-NjotGS84PdUIuV2-2Ur:b76l59m) ([raw file](https://github.com/cockroachdb/cockroach/blob/dacf778374f98e179b850052a580f2d951f1e81b/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L238)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We could use a `util.FastIntMap` to keep this on the stack.
</blockquote></details>

I really don't want to bundle code/allocation optimizations in a behavior change CL that has some urgency in its backport. All existing Oracle implementations allocate slices in their implementations of this method in existing releases, so if we want to come golf allocations in this area I'd prefer to do it later, in standalone changes.
___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 216 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-Njo9vi08l_QpSFssYjP:-Njov8H_BYhqjUPJAJHl:b-gt55np) ([raw file](https://github.com/cockroachdb/cockroach/blob/de574fcf1cf1e340a369c3956190947535054dec/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L216)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"will plan balancing random" sounds off. "randomly"?
</blockquote></details>

Good catch.

Done.
___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 208 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-NjoDaB8Asij1Yvg0X3x:-Njot3Ex9F3IaoNNa5f6:b-vmyp03) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b2af5b4ee5f349f5230f0d849988969f3996eba/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L208)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this live in `pkg/sql/physicalplan/replicaoracle/oracle.go` next to the rest of the `Oracle` implementations?
</blockquote></details>

I don't think so. a) This is a follower read oracle, so I think it belongs here in the followerread package b) we only intend to use it in CCL code (backup and cdc) and our guidance is to keep as much CCL as possible going forward so this seems like an easy way to do so.
___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 245 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/114710#-NjoDzqw9rCTSMrKV_lg:-Njou6qB6UIRDvamZQnB:bhv9d1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b2af5b4ee5f349f5230f0d849988969f3996eba/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L245)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we be returning `true` for `ignoreMisplannedRanges` here and below? For the same reason that `closestOracle.ChoosePreferredReplica` does.
</blockquote></details>

I think so, yep.

Done.

**Comment by dt** [comment]
@nvanbenschoten are your concerns blocking or things that you think could be addressed, if needed, in follow-ups? I was hoping to get this in to get some baking in the nightlies over the holiday weekend.

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
Comment (adityamaru) [inline pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go:698]: can we assert the expected values of ignoreMisplannedRanges in the test below too?

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (adityamaru) [inline pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:248]: did you mean to return ignoreMisplannedRanges here?

**Review by adityamaru — APPROVED** [review]

File: pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go
Comment (dt) [reply pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go:248]: no, its just true here since we picked matching replica to our satisfaction. it is only false if we had an unfulfilled loc filter I think?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114710)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @dt, @miretskiy, @msbutler, and @nvanbenschoten)

___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 248 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/114710#gh-1402685120:-Njsnxpf4O8rvcTpPRtL:b-kvkmhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/24ff8638ee06fbb4ff490d6ee673ac2d104ab62d/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L248)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

no, its just true here since we picked matching replica to our satisfaction. it is only false if we had an unfulfilled loc filter I think?
</blockquote></details>

In short, we should return `ignoreMisplannedRanges = true` in all cases here.

Long answer is that `ignoreMisplannedRanges` is only (at least currently) used when we plan `TableReader`s, and since this new oracle should never be used for that, the returned boolean doesn't really matter. Its goal is to signify to the DistSQL physical planner that the oracle might have chosen a replica that is not the leaseholder. Returning `ignoreMisplannedRanges = true` is the right choice whenever we might be planning on followers.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114710)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @miretskiy, @msbutler, @nvanbenschoten, and @yuzefovich)

___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go` line 248 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/114710#gh-1402685120:-Njt-Imm972CgiKIuX7D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/24ff8638ee06fbb4ff490d6ee673ac2d104ab62d/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L248)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

In short, we should return `ignoreMisplannedRanges = true` in all cases here.

Long answer is that `ignoreMisplannedRanges` is only (at least currently) used when we plan `TableReader`s, and since this new oracle should never be used for that, the returned boolean doesn't really matter. Its goal is to signify to the DistSQL physical planner that the oracle might have chosen a replica that is not the leaseholder. Returning `ignoreMisplannedRanges = true` is the right choice whenever we might be planning on followers.
</blockquote></details>

Done.

**Comment by dt** [comment]
I'm going to go ahead and bors this so weget some nightly runs on master over the long holiday weekend. It'll just be on master for the next week in case there's anything we want to do in follow-ups before we talk about backports.

TFTRs!

bors r=miretskiy,adityamaru

**Pull request merged**
